src/hb-ot-var-common.hh

```diff
@@ -528,6 +584,326 @@ struct tuple_delta_t
 
     return out;
   }
+
+  /* deltas should be compiled already before we compile tuple
+   * variation header cause we need to fill in the size of the
+   * serialized data for this tuple variation */
+  //TODO(qxliu):add option to use sharedTuples in gvar
+  bool compile_tuple_var_header (const hb_map_t& axes_index_map,
+                                 unsigned points_data_length,
+                                 const hb_map_t& axes_old_index_tag_map)
+  {
+    if (!compiled_deltas) return false;
+
+    unsigned cur_axis_count = axes_index_map.get_population ();
+    /* allocate enough memory: 1 peak + 2 intermediate coords + fixed header size */
+    unsigned alloc_len = 3 * cur_axis_count * (F2DOT14::static_size) + 4;
+    char *p = (char *) hb_calloc (alloc_len, sizeof (char));
+    if (unlikely (!p)) return false;
+
+    /* the first 4 bytes are fixed header bytes*/
+    unsigned pos = 4;
+    unsigned flag = 0;
+    F2DOT14* end = reinterpret_cast<F2DOT14 *> (p + alloc_len);
+    /* encode peak coords */
+    F2DOT14* peak_coords_end = encode_peak_coords(p+pos, end, flag, axes_index_map, axes_old_index_tag_map);
+    if (!peak_coords_end)
+    { hb_free (p); return false; }
+
+    F2DOT14* interm_coords_end = encode_interm_coords (peak_coords_end, end, flag, axes_index_map, axes_old_index_tag_map);
+    if (!interm_coords_end)
+    { hb_free (p); return false; }
+
+    //TODO(qxliu): add option to use shared_points in gvar
+    flag |= TupleVariationHeader::TuppleIndex::PrivatePointNumbers;
+
+    unsigned serialized_data_size = points_data_length + compiled_deltas.length;
+    TupleVariationHeader *o = reinterpret_cast<TupleVariationHeader *> (p);
+    o->varDataSize = serialized_data_size;
+    o->tupleIndex = flag;
+
+    unsigned total_header_len = reinterpret_cast<char *> (interm_coords_end) - p;
+    compiled_tuple_header = byte_data_t (p, total_header_len);
+    return true;
+  }
+
+  F2DOT14* encode_peak_coords (char* p, F2DOT14* end,
+                               unsigned& flag,
+                               const hb_map_t& axes_index_map,
+                               const hb_map_t& axes_old_index_tag_map) const
+  {
+    unsigned orig_axis_count = axes_old_index_tag_map.get_population ();
+    F2DOT14* out = reinterpret_cast<F2DOT14 *> (p);
```

Comment:
Instead of passing around and accessing the raw pointers directly you could use hb_array_t<Type> which will provide automatic bounds checking. Since arrays are iterators they also support things like ++, so most of the code interacting with them shouldn't really need to change.